docs: document chatAgents, languageModelChatProviders, and languageModelTools contribution points (#9969)
Microsoft/vscode-docs: api/references/contribution-points.md

```

## contributes.chatAgents

Contributes [custom agents](/docs/agent-customization/custom-agents.md) for Copilot Chat. Custom agents are pre-configured AI personas with specific instructions and tool restrictions. Use this contribution point to bundle reusable custom agents with your extension so they appear alongside user-defined agents in the agents dropdown.

Each entry requires a `path` to a `.agent.md` file relative to the extension root. You can optionally specify a `when` clause to conditionally enable the agent. Specify the agent's `name`, `description`, and other metadata inside the `.agent.md` frontmatter rather than in the contribution point.

```json
{
  "contributes": {
    "chatAgents": [
      {
        "path": "./agents/planner.agent.md"
      }
    ]
  }
}
```

### chatAgents properties

| Property | Type | Required | Description |
| -------- | ---- | -------- | ----------- |
| `path` | `string` | Yes | Path to the `.agent.md` file relative to the extension root. The path must resolve to a location inside the extension. |
| `when` | `string` | No | A [when clause](/api/references/when-clause-contexts) condition that must be true for this entry to be enabled. |

> 

Reviewer: ```suggestion
| `sessionTypes` | `string[]` | No | The chat session types where this agent should be offered. |

```

}
```

## contributes.languageModelChatProviders

Contributes a language model chat provider to VS Code, enabling extensions to supply custom language models that users can select in the model picker. Each provider manages its own set of models and handles chat requests on their behalf.

Register one entry per provider, giving it a unique `vendor` ID. Then use `vscode.lm.registerLanguageModelChatProvider` in your extension activation to wire up the implementation.

```json
{
  "contributes": {
    "languageModelChatProviders": [
      {
        "vendor": "my-provider",
        "displayName": "My Provider"
      }
    ]
  }
}
```

> You can add an optional `managementCommand` to let users configure the provider (for example, to enter API keys). The value must be a command declared in `contributes.commands`:

Reviewer: ```suggestion
To let users configure the provider (for example, to enter API keys), add a `configuration` schema. Mark sensitive fields with `"secret": true` so they are stored securely:
```

```json
{
  "contributes": {
    "languageModelChatProviders": [
      {
        "vendor": "my-provider",
        "displayName": "My Provider",
        "managementCommand": "my-provider.manage"
      }
    ],
    "commands": [
      {
        "command": "my-provider.manage",
        "title": "Manage My Provider"
      }
    ]
  }
}

> ```

Reviewer: ```suggestion
```json
{
  "contributes": {
    "languageModelChatProviders": [
      {
        "vendor": "my-provider",
        "displayName": "My Provider",
        "configuration": {
          "type": "object",
          "properties": {
            "apiKey": {
              "type": "string",
              "secret": true
            }
          }
        }
      }
    ]
  }
}
```

### languageModelChatProviders properties

| Property | Type | Required | Description |
| -------- | ---- | -------- | ----------- |
| `vendor` | `string` | Yes | Unique identifier for the provider, used as the first argument to `vscode.lm.registerLanguageModelChatProvider`. |
| `displayName` | `string` | Yes | Human-readable name shown in the model picker UI. |

> | `managementCommand` | `string` | No | Command ID that opens a UI for managing this provider (for example, configuring API keys). Must be declared in `contributes.commands`. |

Reviewer: ```suggestion
| `configuration` | `object` | No | A JSON schema describing configuration options for the provider (for example, API keys). Properties can be marked `"secret": true` to store them securely. This is the recommended way to let users configure a provider. |
| `managementCommand` | `string` | No | _Deprecated. Use `configuration` instead._ Command ID that opens a UI for managing this provider. Must be declared in `contributes.commands`. |
| `when` | `string` | No | A [when clause](/api/references/when-clause-contexts) that controls whether this provider appears in the Manage Models list. |
```

```json
{
  "contributes": {
    "languageModelTools": [
      {
        "name": "my-extension_queryDatabase",
        "displayName": "Query Database",
        "modelDescription": "Executes a read-only SQL query against the project database and returns the results as JSON. Use this tool when the user asks about data stored in the database.",
        "canBeReferencedInPrompt": true,
        "toolReferenceName": "queryDatabase",
        "icon": "$(database)",
        "inputSchema": {
          "type": "object",
          "properties": {
            "query": {
              "type": "string",
              "description": "The SQL SELECT statement to execute."
            }
          },
          "required": ["query"]
        }
      }
    ]
  }
}
```

### languageModelTools properties

| Property | Type | Required | Description |
| -------- | ---- | -------- | ----------- |
| `name` | `string` | Yes | Unique name of the tool used in the extension implementation. Use the `{verb}_{noun}` format and prefix with your extension name to avoid collisions. |

> | `displayName` | `string` | No | User-friendly name displayed in the UI. |

Reviewer: ```suggestion
| `displayName` | `string` | Yes | User-friendly name displayed in the UI. |
```
